[2022.10.18] Feat: pptx parser 유틸 추가

0jaaack opened the pull request:
### pptx parser 유틸 추가

### Task
[Task Card Notion Link](https://www.notion.so/vanillacoding/parse-pptx-view-data-2a59e5abbfff460c9f9621fc70618709)

### Description
- pptx 파일을 읽어 view data로 추출해주는 유틸 함수를 생성하였습니다.
- 해당 view data 기본 구조는,
```
{
  slideWidth: (number),
  slideHeight: (number),
  slides: (array)
}
```
의 구조로 구성되어 있습니다.
- slides의 내부는 각 슬라이드들에 대한 정보가 담겨있으며,
```
{
  slideId: (string),
  items: (array)
}...
```
의 구조로 구성되어 있습니다.
- items의 내부는 각 요소에 대한 정보가 담겨있으며,
```
{
  type: (string),
  order: (number),
  id: {
      itemId: (string),
      itemName: (string)
  },
  width: (number),
  height: (number),
  x: (number),
  y: (number),
  content: (object),
}
```
의 구성으로 이루어져있습니다.
- content object 내부 구성은, 요소의 type에 따라 달라집니다.
- 요소의 type이 text일 경우, 다음과 같은 내부 구성을 가집니다.
```
{
  value: (string),
  isBold: (boolean),
  isItalic: (boolean),
  isUnderlined: (boolean),
  fontColor: (string)
  size: (number),
  backgroundColor: (string),
}
```
- 요소의 type이 image일 경우, 다음과 같은 내부 구성을 가집니다.
```
{
  src: (string)
}
```
- view render 로직 및 diffing 로직, 파일 병합 로직에 사용될 데이터의 소스로 쓰일 데이터입니다.

### Point
- 기본적으로 .pptx 파일은, slide와 slide 관련 정보들을 담고 있는 .xml 파일들과 미디어 파일들의 폴더가 압축되어있는 구조입니다.
- 그래서 xml 파일을 읽기 위해서, Javascript의 DOMParser를 사용해 파일을 읽어 view data를 가져오게 되었습니다.
- 원래는 서버에서 parser를 진행하려고 하였으나, DOMParser를 브라우저 환경에서만 사용할 수 있어 부득이하게 client 상에서 실행하게 되었습니다.
- nodejs 환경에서 DOMParser를 대체할 수 있는 라이브러리 등을 찾아보았지만, 기능적으로 부족한 점이 많아(ex_querySelector를 이용한 중첩된 자식 선택자 등) 사용하지 않았습니다.
- 해당 로직을 사용하기 위해서 파일을 압축해제하여 읽을 수 있는, `JSZip` 라이브러리를 사용하였습니다.
- config 폴더에, 미리 정의된 color 값을 읽기 위한 preset 값이 정의되어 있습니다.
- 코드의 양이 조금 많아 혹시 제가 실수하거나 놓친 부분이 없는 지 잘 살펴주시길 바라겠습니다.!

### ETC
[기술 구현 검증 시 Notion 문서](https://www.notion.so/vanillacoding/pptx-view-data-0b5fa674515c4256a5a6f804dca4cf6a)

0jaaack committed:
Feat: pptx parser 유틸 추가
1. pptx 파일을 읽어 view data로 추출해주는 유틸 함수를 생성하였습니다.
2. 해당 유틸 사용위해, jszip 라이브러리 사용하였습니다.
3. slide의 기본 크기/각 slide별 id/slide 내 요소들에 대한 정보를 가져오는 로직입니다.
4. 요소는 현재 image와 text를 지원하는 상태입니다.

Y-SeonJong commented on src/config/constants/presetColors.js:
```diff
@@ -0,0 +1,84 @@
+  },
+};
+
+module.exports = PRESET_COLORS;
```
👍

Donghyeon0908 commented on src/utils/pptxParser.js:
```diff
@@ -0,0 +1,219 @@
+import JSZip from "jszip";
+import PRESET_COLORS from "../config/constants/presetColors";
+import ITEM_TYPES from "../config/constants/itemTypes";
+
+const parser = new DOMParser();
+const EMUFactor = 96 / 914400;
+
```
리뷰 시작하겠습니다 👍

Donghyeon0908 reviewed (commented):
재혁님이 작성해주신 로직을 복사해서 돌려봤는데 .pptx파일마다 오류가 나는곳이 다르다보니 값들이 제대로 읽혀오지않은듯해요ㅠ
현재 44, 65, 108번째 줄에서 오류가 발생하는것을 확인했으니 검토부탁드립니다..!

Donghyeon0908 commented on src/utils/pptxParser.js:
```diff
@@ -0,0 +1,219 @@
+import ITEM_TYPES from "../config/constants/itemTypes";
+
+const parser = new DOMParser();
+const EMUFactor = 96 / 914400;
```
사이즈를 픽셀로 변환할 값을 구하셨군요 👍

Donghyeon0908 commented on src/utils/pptxParser.js:
```diff
@@ -0,0 +1,219 @@
+};
+
+const getSlidePaths = async (pptx) => {
+  const contentTypes = await pptx.file("[Content_Types].xml").async("text");
+  const slideOverrides = parser
+    .parseFromString(contentTypes, "application/xml")
+    .querySelectorAll("Override");
+
+  return Array.from(slideOverrides)
+    .filter((el) =>
+      el.attributes.PartName.nodeValue.includes("/ppt/slides/slide"),
+    )
+    .map((el) => el.attributes.PartName.nodeValue.slice(1));
```
저 이 로직이 약간 이해가 안되서 그러는데 설명 부탁드릴게요~ slice(1)을 하면 Path값을 구할수있나요? 🥸

0jaaack replied:
넵 문자열의 첫 글자를 쓰지 않기 위해서 배열의 `map()` 메소드를 사용해, 각 path 문자열마다 `slice(1)`을 적용하였습니다~ 

https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/String/slice

Donghyeon0908 replied:
아~ 첫글자를 배제하셨군요 감사합니다!

Donghyeon0908 commented on src/utils/pptxParser.js:
```diff
@@ -0,0 +1,219 @@
+  );
+  const imagePath = embedIdRelXmlDOM.attributes.Target.nodeValue;
+  const imgSrc = await pptx
+    .file(imagePath.replace("..", "ppt"))
```
로직자체는 어떻게 돌아가는지 이해는 되는데 그래도 src 주소를 얻는게 신기하네요🧐

Donghyeon0908 reviewed (approved):
재혁님 드디어 파싱이 완성되었군요...! 지난주 기술검증부터 정말 고생많으셨습니다 저도 코드를 보고 `Promise.all` 이나 옵셔널체이닝, nullish 병합 연산자를 잘 활용하시는 모습을 보면서 다시 한번 배우고 있네요! 
 이제 구현에 집중하면 되니 조금만 더 화이팅합시다! 수고하셨어요~

Y-SeonJong commented on src/utils/pptxParser.js:
```diff
@@ -0,0 +1,219 @@
+  return Array.from(slideOverrides)
+    .filter((el) =>
+      el.attributes.PartName.nodeValue.includes("/ppt/slides/slide"),
+    )
```
ppt/slides/slide 가 핵심 파일이기 때문에 ppt/slides/slide라는 단어가 있는걸로만 filter 하는 군요!

Y-SeonJong commented on src/utils/pptxParser.js:
```diff
@@ -0,0 +1,219 @@
+  const IdListXml = slide.querySelectorAll("cSld extLst");
+  const slideIdXml = Array.from(IdListXml).find(
+    ({ nodeName }) => nodeName === "p:extLst",
+  );
```
find와 구조분해 할당의 적절항 활용..!!!

Y-SeonJong commented on src/utils/pptxParser.js:
```diff
@@ -0,0 +1,219 @@
+  const presetFontColor = presetFontXml?.attributes.val.nodeValue;
+  const presetFontDarkness =
+    textAttributeXml.querySelector("lumMod")?.attributes.val.nodeValue;
+
```
optional Chaining ( ?. ) 이란?
Error 를 보내는 대신 undefined로 대체 한다.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

Y-SeonJong commented on src/utils/pptxParser.js:
```diff
@@ -0,0 +1,219 @@
+  return {
+    value,
+    isBold: !!b?.nodeValue,
+    isItalic: !!i?.nodeValue,
```
느낌표 두개(!!) 는 undefined, "", 0 값은 false. 그외에는 true 값을 리턴
확인했습니다!

Y-SeonJong commented on src/utils/pptxParser.js:
```diff
@@ -0,0 +1,219 @@
+  return Promise.all(
+    items.map(async (item, index) => {
+      const type = ITEM_TYPES[item.tagName] ?? "unidentified item";
+      const id = getItemId(item);
```
값이 null 이나 undefined일 때 
Nullish coalescing operator 인 ?? 를 사용해서 
기본 값을 unidentifed item으로 주셨네요 O_O!

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

Y-SeonJong reviewed (approved):
와.. 진짜 거대한 산 중 하나가 pptx 파일을 다 가져오는 거였는데.. 정말 세세하게 작성하셨네요..!
보면서 코드적으로도 배울게 많은 것 같아요.!
구조 분해 할당도 이 코드 하나면 누구나 알지 않을까!
특히
!!
?.
??
등 자주 써보지 못한 표현들에 대해서 공부를 할 수 있게 하는 좋은 기회였어요~!

0jaaack committed:
Fix: 엣지 케이스 대응 추가
1. JS에서 커스텀하여 .pptx 파일 생성한 경우 id값 대응 케이스 추가
2. ppt 자체 기본 텍스트 박스에서 내용 작성 시, 위치 및 크기 값 대응 케이스 추가
3. font color hex code 변환 시, 앞에 #이 붙도록 추가.
